lib/recursive-methods.rb

```diff
@@ -1,49 +1,87 @@
+require 'pry'
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
-
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError.new("BAD") if n < 0 
+  return 1 if n == 0
+  return n if n == 1
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  return reverse(s[1..-1]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def reverse_inplace(s, i = 0)
+  return s if i == s.length / 2
+  s[i], s[s.length - i - 1] = s[s.length - i - 1], s[i]
+  s = reverse_inplace(s, i += 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n <= 0
+  return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2) since using .length?
+# Space complexity: O(n)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  return false if s.length % 2 != 0
+  return true if s == ""
+  return true if s == "()"
+  center_l = s.length / 2 - 1
+  center_r = center_l + 1
+  if s[center_l] + s[center_r] != "()"
+    return false
+  else
+    next_string = s[center_l - 1] + s[center_r + 1]
```

Comment:
This means you're going from a string which could be 14 characters long immediately to a string of 2 characters...

This won't work in all cases.